### PR TITLE
Add Thrift support for for SqlInvokedFunction and SqlFunctionId

### PR DESCRIPTION
**Pull request by SeanIFitch**
## Motivation and Context
`SessionRepresentation` has Thrift annotations but some 
of the classes of its member variables do not have Thrift support.
Add annotations to `SqlInvokedFunction` and `SqlFunctionId`.
Work towards Thrift serialization support of `TaskResource`.
https://github.com/prestodb/presto/issues/19839

## Testing
Add serde test for new `TypeSignatureParameterUnion`

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java [hunk 1]
```diff
@@ -91,18 +102,32 @@ public boolean hasTypeName()
         return typeName.isPresent();
     }
 
+    //@ThriftField(1)
```
Comment (ajaygeorge) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java:105]: nit : remove comment

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java [hunk 2]
```diff
@@ -91,18 +102,32 @@ public boolean hasTypeName()
...
     public QualifiedObjectName getTypeName()
     {
         checkArgument(typeName.isPresent(), "TypeSignatureBase %s does not have type name", toString());
         return typeName.get();
     }
 
+    //@ThriftField(2)
```
Comment (ajaygeorge) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureBase.java:112]: nit : remove comment

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureParameter.java [hunk 3]
```diff
@@ -13,19 +13,25 @@
...
+import static com.facebook.presto.common.type.TypeSignatureParameterUnion.convertToValue;
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 
+@ThriftStruct
 public class TypeSignatureParameter
 {
     private final ParameterKind kind;
-    private final Object value;
+    private final TypeSignatureParameterUnion value;
```
Comment (ajaygeorge) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureParameter.java:34]: does this work with the previous JSON serde?

**Review by ajaygeorge — COMMENTED** [review]
LGTM % TaskUpdateRequest changes.
Curious why have only partially added support. 
Maybe we can move it out of this PR and do it in a separate one.

File: presto-main/src/main/java/com/facebook/presto/server/TaskUpdateRequest.java [hunk 4]
```diff
@@ -64,20 +69,33 @@ public TaskUpdateRequest(
...
 
+    //This constructor is for testing purposes only and will soon be deprecated
+    @ThriftConstructor
+    public TaskUpdateRequest(
+            @JsonProperty("session") SessionRepresentation session,
+            @JsonProperty("extraCredentials") Map<String, String> extraCredentials,
+            @JsonProperty("fragment") Optional<byte[]> fragment)
+    {
+        this(session, extraCredentials, fragment, ImmutableList.of(), OutputBuffers.createInitialEmptyOutputBuffers(ARBITRARY), Optional.empty());
+    }
```
Comment (ajaygeorge) [inline presto-main/src/main/java/com/facebook/presto/server/TaskUpdateRequest.java:80]: Curious why have only partially added support. Can we move this out into a separate PR / commit.

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureParameter.java (hunk 3 above)
Comment (SeanIFitch) [reply presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureParameter.java:34]: Yes; JSON uses the `toString()` method, so I simply cast the `TypeSignatureParameterUnion` back to an `Object` in that method.

File: presto-main/src/main/java/com/facebook/presto/server/TaskUpdateRequest.java (hunk 4 above)
Comment (SeanIFitch) [reply presto-main/src/main/java/com/facebook/presto/server/TaskUpdateRequest.java]: Removed it for now.

**Review by ajaygeorge — APPROVED** [review]
LGTM

**Review by ajaygeorge — COMMENTED** [review]
One thing I missed is , since we are introducing a union type, it will be good to have a test that covers the serde of that code path.

File: presto-common/pom.xml [hunk 5]
```diff
@@ -102,5 +102,15 @@
...
+        <dependency>
+            <groupId>com.facebook.drift</groupId>
+            <artifactId>drift-codec</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (ajaygeorge) [inline presto-common/pom.xml:114]: Is logback-classic needed as a test dependency?

**Comment by ajaygeorge** [comment]
@SeanIFitch please check the maven failures. I think they are related to the latest changes.

File: presto-common/pom.xml [hunk 6]
```diff
@@ -102,5 +102,17 @@
...
+
+        <dependency>
+            <groupId>com.facebook.drift</groupId>
+            <artifactId>drift-protocol</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
-</project>
+</project>
```
Comment (ajaygeorge) [inline presto-common/pom.xml:118]: stray new line?

**Review by ajaygeorge — APPROVED** [review]
LGTM % a tiny nit.
Also can you double check if the commit message is still accurate. If not please fix that as well.

File: presto-common/pom.xml (hunk 6 above)
Comment (SeanIFitch) [reply presto-common/pom.xml]: For some reason adding this new line causes it to fail Maven tests for check-other-modules… any idea what might cause this?

Comment (ajaygeorge) [reply presto-common/pom.xml]: I think the error is not related. Can you try rebasing with master once. We had several flaky tests in the past. Maybe this has been fixed in some of the recent commits.

**Comment by ajaygeorge** [comment]
@SeanIFitch you need to rebase with master. I think you have an extra merge commit now. Can you get rid of that and do a rebase and push.

**Comment by ajaygeorge** [comment]
@SeanIFitch Looks like all the tests are passing now.
Can you double check the commit message.
I don't see `SessionRepresentation` in the list of changed files. Is the commit message still accurate?

**Comment by SeanIFitch** [comment]
> @SeanIFitch Looks like all the tests are passing now. Can you double check the commit message. I don't see `SessionRepresentation` in the list of changed files. Is the commit message still accurate?

The purpose of this is to fix some of the member variable classes of `SessionRepresentation` not being supported for thrift serialization. I updated the commit message to specify those classes.

**Comment by ajaygeorge** [comment]
@SeanIFitch Let us reword the commit message as
`Add thrift support for SqlInvokedFunction and SqlFunctionId`

**Review by tdcmeehan — APPROVED** [review]
